--- src/org/kapott/hbci/GV_Result/GVRKUms.java	2010-04-12 11:02:59.000000000 +0200
+++ src/org/kapott/hbci/GV_Result/GVRKUms.java	2010-10-07 15:48:35.369618105 +0200
@@ -664,7 +664,7 @@
                     }
 
                     // set default values for optional non-given bdates
-                    if (btag.start.timestamp==null) {
+                    if (btag.start != null && btag.start.timestamp==null) {
                         btag.start.timestamp=btag.end.timestamp;
                     }
                     for (Iterator j=btag.lines.iterator(); j.hasNext(); ) {
