Index: src/org/kapott/hbci/passport/AbstractHBCIPassport.java
===================================================================
--- src/org/kapott/hbci/passport/AbstractHBCIPassport.java	(revision 439)
+++ src/org/kapott/hbci/passport/AbstractHBCIPassport.java	(revision 440)
@@ -245,6 +245,7 @@
                 entry.blz=upd.getProperty(header+".KTV.KIK.blz");
                 entry.country=upd.getProperty(header+".KTV.KIK.country");
                 entry.number=number;
+                entry.subnumber=upd.getProperty(header+".KTV.subnumber");
                 entry.curr=upd.getProperty(header+".cur");
                 entry.type=upd.getProperty(header+".konto");
                 entry.customerid=upd.getProperty(header+".customerid");
