Index: src/org/kapott/hbci/passport/AbstractHBCIPassport.java
===================================================================
--- src/org/kapott/hbci/passport/AbstractHBCIPassport.java	2011-03-11 18:08:28.184000253 +0100
+++ src/org/kapott/hbci/passport/AbstractHBCIPassport.java	2011-03-11 18:08:52.908000253 +0100
@@ -422,7 +422,7 @@
 
     public final int getMaxMsgSizeKB()
     {
-        return (bpd!=null)?Integer.parseInt(bpd.getProperty("BPA.maxmsgsize")):0;
+        return (bpd!=null)?Integer.parseInt(bpd.getProperty("BPA.maxmsgsize","0")):0;
     }
 
     public final String[] getSuppLangs()
