Index: src/challengedata.xml
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/challengedata.xml,v
retrieving revision 1.4
diff -u -r1.4 challengedata.xml
--- src/challengedata.xml	17 May 2011 16:39:07 -0000	1.4
+++ src/challengedata.xml	18 May 2011 17:18:17 -0000
@@ -676,7 +676,7 @@
       <param type="value">BTG.value</param>
       <param>Other.KIK.blz</param>
       <param>Other.number</param>
-      <param>date</param>
+      <param type="date">date</param>
     </challengeinfo>
   </job>
 
Index: src/org/kapott/hbci/manager/ChallengeInfo.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/manager/ChallengeInfo.java,v
retrieving revision 1.4
diff -u -r1.4 ChallengeInfo.java
--- src/org/kapott/hbci/manager/ChallengeInfo.java	17 May 2011 16:39:07 -0000	1.4
+++ src/org/kapott/hbci/manager/ChallengeInfo.java	18 May 2011 17:18:17 -0000
@@ -21,7 +21,9 @@
 package org.kapott.hbci.manager;
 
 import java.io.InputStream;
+import java.text.SimpleDateFormat;
 import java.util.ArrayList;
+import java.util.Date;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -412,9 +414,34 @@
       if (value == null || value.length() == 0)
         return value;
 
-      // Ist es ein Betrag?
-      if (type != null && type.equals("value"))
-        return new SyntaxWrt(value, 1, 0).toString(0);
+      // haben wir einen konkreten Typ?
+      if (type != null && type.length() > 0)
+      {
+        // Ist es ein Betrag?
+        if (type.equals("value"))
+        {
+          return new SyntaxWrt(value, 1, 0).toString(0);
+        }
+        // Datum?
+        else if (type.equals("date"))
+        {
+          try
+          {
+            // War ja klar, dass die das Datum hier nochmal in einem anderen Format haben wollen
+            Date date = new SimpleDateFormat("yyyy-MM-dd").parse(value);
+            return new SimpleDateFormat("ddMMyyyy").format(date);
+          }
+          catch (Exception e)
+          {
+            HBCIUtils.log("unable to parse " + value + " as yyyy-MM-dd, sending as is",HBCIUtils.LOG_WARN);
+            return value;
+          }
+        }
+        else
+        {
+          HBCIUtils.log("unknown parameter type " + type,HBCIUtils.LOG_WARN);
+        }
+      }
       
       // Ansonsten ganz normal den Betrag zurueckliefern
       return value;
