Index: src/org/kapott/hbci/GV/HBCIJobImpl.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/GV/HBCIJobImpl.java,v
retrieving revision 1.2
diff -u -r1.2 HBCIJobImpl.java
--- src/org/kapott/hbci/GV/HBCIJobImpl.java	17 May 2011 16:39:06 -0000	1.2
+++ src/org/kapott/hbci/GV/HBCIJobImpl.java	23 May 2011 09:43:15 -0000
@@ -56,7 +56,8 @@
 public abstract class HBCIJobImpl 
     implements HBCIJob
 {
-    private String name;              /* @brief name of the corresponding GV-segment */
+    private String name;              /* Job-Name mit Versionsnummer */
+    private String jobName;           /* Job-Name ohne Versionsnummer */
     private String segVersion;        /* Segment-Version */
     private Properties llParams;       /* Eingabeparameter für diesen GV (Saldo.KTV.number) */
     private HBCIPassportList passports;
@@ -112,7 +113,7 @@
     {
         StringBuffer ret=null;
         
-        // entfernen der versionsnummer vom aktuellen jobnamen
+        // Macht aus z.Bsp. "KUmsZeit5" -> "KUmsZeitPar5.SegHead.code"
         StringBuffer searchString=new StringBuffer(name);
         for (int i=searchString.length()-1;i>=0;i--) {
             if (!(searchString.charAt(i)>='0' && searchString.charAt(i)<='9')) {
@@ -194,9 +195,49 @@
             throw new JobNotSupportedException(jobnameLL);
         }
         
-        // namen+versionsnummer zurückgeben
+        // namen+versionsnummer speichern
+        this.jobName    = jobnameLL;
         this.segVersion = Integer.toString(maxVersion);
-        this.name = jobnameLL + this.segVersion;
+        this.name       = jobnameLL + this.segVersion;
+    }
+    
+    /**
+     * Legt die Versionsnummer des Segments manuell fest.
+     * Ist u.a. noetig, um HKTAN-Segmente in genau der Version zu senden, in der
+     * auch die HITANS empfangen wurden. Andernfalls koennte es passieren, dass
+     * wir ein HKTAN mit einem TAN-Verfahren senden, welches in dieser HKTAN-Version
+     * gar nicht von der Bank unterstuetzt wird. Das ist ein Dirty-Hack, ich weiss ;)
+     * Falls das noch IRGENDWO anders verwendet wird, muss man hoellisch aufpassen,
+     * dass alle Stellen, wo "this.name" bzw. "this.segVersion" direkt oder indirekt
+     * verwendet wurde, ebenfalls beruecksichtigt werden.
+     * @param version die neue Versionsnummer.
+     */
+    public void setSegVersion(String version)
+    {
+      if (version == null || version.length() == 0)
+      {
+        HBCIUtils.log("tried to change segment version for task " + this.jobName + " explicit, but no version given",HBCIUtils.LOG_WARN);
+        return;
+      }
+      
+      // Wenn sich die Versionsnummer nicht geaendert hat, muessen wir die
+      // Huehner ja nicht verrueckt machen ;)
+      if (version.equals(this.segVersion))
+        return;
+      
+      HBCIUtils.log("changed segment version for task " + this.jobName + " explicit to " + version,HBCIUtils.LOG_INFO);
+
+      // Bisheriges Request-Tag entfernen
+      this.llParams.remove(this.name);
+
+      // Neue Versionsnummer speichern
+      this.segVersion = version;
+      
+      // Neuer Name des Jobs
+      this.name = this.jobName + this.segVersion;
+      
+      // Request-Tag neu setzen
+      this.llParams.setProperty(this.name,"requested");
     }
     
     public int getMaxNumberPerMsg()
Index: src/org/kapott/hbci/passport/AbstractPinTanPassport.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/AbstractPinTanPassport.java,v
retrieving revision 1.4
diff -u -r1.4 AbstractPinTanPassport.java
--- src/org/kapott/hbci/passport/AbstractPinTanPassport.java	17 May 2011 16:39:07 -0000	1.4
+++ src/org/kapott/hbci/passport/AbstractPinTanPassport.java	23 May 2011 09:43:15 -0000
@@ -925,9 +925,10 @@
             
             HBCIUtils.log("afterCustomDialogInitHook: patching message queues for twostep method",HBCIUtils.LOG_DEBUG);
             
-            HBCIHandler handler=(HBCIHandler)getParentHandlerData();
-            Properties  secmechInfo=getCurrentSecMechInfo();
-            String      process=secmechInfo.getProperty("process");
+            HBCIHandler handler     = (HBCIHandler)getParentHandlerData();
+            Properties  secmechInfo = getCurrentSecMechInfo();
+            String      segversion  = secmechInfo.getProperty("segversion");
+            String      process     = secmechInfo.getProperty("process");
             
             List msgs=dialog.getMessages();
             List new_msgs=new ArrayList();
@@ -955,7 +956,12 @@
                             // neue msg erzeugen
                             additional_msg_tasks=new ArrayList();
 
-                            HBCIJob hktan=handler.newJob("TAN2Step");
+                            GVTAN2Step hktan = (GVTAN2Step) handler.newJob("TAN2Step");
+                            
+                            // muessen wir explizit setzen, damit wir das HKTAN in der gleichen Version
+                            // schicken, in der das HITANS kam.
+                            hktan.setSegVersion(segversion);
+                            
                             hktan.setParam("process",process);
                             hktan.setParam("notlasttan","N");
                             
@@ -1028,7 +1034,7 @@
                             }
 
                             // willuhn 2011-05-09: Bei Bedarf noch das TAN-Medium erfragen
-                            applyTanMedia((GVTAN2Step) hktan);
+                            applyTanMedia(hktan);
                             
                             // hktan-job zur neuen msg hinzufügen
                             additional_msg_tasks.add(hktan);
@@ -1049,7 +1055,12 @@
                             new_msg_tasks.add(task);
                             
                             // dazu noch einen hktan-job hinzufügen
-                            GVTAN2Step hktan1=(GVTAN2Step)handler.newJob("TAN2Step");
+                            GVTAN2Step hktan1 = (GVTAN2Step) handler.newJob("TAN2Step");
+
+                            // muessen wir explizit setzen, damit wir das HKTAN in der gleichen Version
+                            // schicken, in der das HITANS kam.
+                            hktan1.setSegVersion(segversion);
+
                             hktan1.setParam("process","4");
                             // TODO: evtl. listindex ermitteln
                             // hktan1.setParam("listidx","");
@@ -1067,7 +1078,12 @@
                             additional_msg_tasks=new ArrayList();
                             
                             // HKTAN-job für das einreichen der TAN erzeugen
-                            GVTAN2Step hktan2=(GVTAN2Step)handler.newJob("TAN2Step");
+                            GVTAN2Step hktan2 = (GVTAN2Step) handler.newJob("TAN2Step");
+
+                            // muessen wir explizit setzen, damit wir das HKTAN in der gleichen Version
+                            // schicken, in der das HITANS kam.
+                            hktan1.setSegVersion(segversion);
+
                             hktan2.setParam("process","2");
                             hktan2.setParam("notlasttan","N");
                             // TODO: evtl. listindex ermitteln
