Index: src/org/kapott/hbci/manager/ChallengeInfo.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/manager/ChallengeInfo.java,v
retrieving revision 1.8
diff -u -r1.8 ChallengeInfo.java
--- src/org/kapott/hbci/manager/ChallengeInfo.java	20 May 2011 11:27:03 -0000	1.8
+++ src/org/kapott/hbci/manager/ChallengeInfo.java	30 May 2011 12:47:19 -0000
@@ -142,12 +142,38 @@
    */
   private String getVersion(Properties secmech)
   {
-    // HHD-Version ermitteln
+    // Das ist die "Technische Kennung"
+    // Siehe "Belegungsrichtlinien TANve1.4  mit Erratum 1-3 final version vom 2010-11-12.pdf"
+    // Der Name ist standardisiert, wenn er mit "HHD1...." beginnt, ist
+    // das die HHD-Version
     String id = secmech.getProperty("id","");
-    
     if (id.startsWith("HHD1.4")) return VERSION_HHD_1_4;
     if (id.startsWith("HHD1.3")) return VERSION_HHD_1_3;
     
+    // Fallback 1. Wir schauen noch in "ZKA-Version bei HKTAN"
+    String version = secmech.getProperty("zkamethod_version");
+    if (version != null && version.length() > 0)
+    {
+      if (version.startsWith("1.4")) return VERSION_HHD_1_4;
+      if (version.startsWith("1.3")) return VERSION_HHD_1_3;
+    }
+    
+    // Fallback 2. Wir checken noch die HITAN/HKTAN-Version
+    // Bei HKTAN5 kann es HHD 1.3 oder 1.4 sein, bei HKTAN4 bleibt eigentlich nur noch 1.3
+    // Ich weiss nicht, ob Fallback 2 ueberhaupt notwendig ist. Denn angeblich
+    // ist zkamethod_version seit HHD 1.3.1 Pflicht (siehe
+    // FinTS_3.0_Security_Sicherheitsverfahren_PINTAN_Rel_20101027_final_version.pdf,
+    // Data dictionary "Version ZKA-TAN-Verfahren"
+    String segversion = secmech.getProperty("segversion");
+    if (segversion != null && segversion.length() > 0)
+    {
+      int i = Integer.parseInt(segversion);
+      if (i == 5)
+        return VERSION_HHD_1_4; // Genau wissen wir es nicht, aber HHD 1.4 ist wahrscheinlich
+      if (i == 4)
+        return VERSION_HHD_1_3; // 1.4 ist in HKTAN4 noch nicht erlaubt, damit bleibt eigentlich nur 1.3
+    }
+    
     // Default:
     return VERSION_HHD_1_2;
   }
Index: src/org/kapott/hbci/passport/HBCIPassportPinTan.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/HBCIPassportPinTan.java,v
retrieving revision 1.4
diff -u -r1.4 HBCIPassportPinTan.java
--- src/org/kapott/hbci/passport/HBCIPassportPinTan.java	27 May 2011 15:46:13 -0000	1.4
+++ src/org/kapott/hbci/passport/HBCIPassportPinTan.java	30 May 2011 12:47:19 -0000
@@ -361,7 +361,7 @@
                     StringBuffer s = flicker != null ? new StringBuffer(flicker) : new StringBuffer();
                     HBCIUtilsInternal.getCallback().callback(this,
                         HBCICallback.NEED_PT_TAN,
-                        secmechInfo.getProperty("name")+"\n"+secmechInfo.getProperty("inputinfo")+"\n"+challenge,
+                        secmechInfo.getProperty("name")+"\n"+secmechInfo.getProperty("inputinfo")+"\n\n"+challenge,
                         HBCICallback.TYPE_TEXT,
                         s);
                     if (s.length()==0) {
