Index: src/org/kapott/hbci/manager/FlickerRenderer.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/manager/FlickerRenderer.java,v
retrieving revision 1.2
diff -u -r1.2 FlickerRenderer.java
--- src/org/kapott/hbci/manager/FlickerRenderer.java	27 May 2011 15:46:13 -0000	1.2
+++ src/org/kapott/hbci/manager/FlickerRenderer.java	6 Jun 2011 15:24:50 -0000
@@ -196,7 +196,16 @@
             // Warten
             // Wir errechnen die Wartezeit in jedem Durchlauf.
             // Dann kann die Frequenz auch waehrend des Blinkens geaendert werden.
-            long sleep = 1000L / freq;
+            // In der Spec. ist nicht eindeutig definiert, ob mit derm Maximalfrequenz
+            // von 20 Hz die Anzahl der Blink-Vorgaenge oder die Anzahl der Zeichen
+            // gemeint ist (wir uebertragen ja jedes Zeichen doppelt - einmal mit
+            // schwarzem Sync und einmal mit weissen). User berichteten mir, dass
+            // die Flicker-Codes auf den Bank-Webseiten schneller blinken als die
+            // von HBCI4Java. Ich halbiere daher die Wartezeit nochmal, weil davon
+            // ausgegangen werden kann, dass mit der Frequenz die Anzahl der Zeichen
+            // gemeint ist und nicht die der Blink-Vorgaenge. Ergo kann eventuell
+            // doppelt so schnell geblinkt werden, wie bisher vorgesehen.
+            long sleep = 1000L / freq / 2;
             sleep(sleep);
           }
         }
