Index: src/org/kapott/hbci/manager/HBCIHandler.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/manager/HBCIHandler.java,v
retrieving revision 1.1
diff -u -r1.1 HBCIHandler.java
--- src/org/kapott/hbci/manager/HBCIHandler.java	4 May 2011 22:37:46 -0000	1.1
+++ src/org/kapott/hbci/manager/HBCIHandler.java	31 Aug 2011 14:03:33 -0000
@@ -156,6 +156,13 @@
      * BIC/IBAN geliefert */
     private void updateSEPAInfo()
     {
+        Properties bpd = passport.getBPD();
+        if (bpd == null)
+        {
+          HBCIUtils.log("have no bpd, skipping SEPA information fetching", HBCIUtils.LOG_WARN);
+          return;
+        }
+
         // jetzt noch zusaetzliche die SEPA-Informationen abholen
         try {
         	if (getSupportedLowlevelJobs().getProperty("SEPAInfo")!=null) {
@@ -183,8 +190,14 @@
         	} else {
         		HBCIUtils.log("institute does not support SEPA accounts, so we skip fetching information about SEPA", HBCIUtils.LOG_DEBUG);
         	}
-        } catch (Exception e) {
-        	throw new HBCI_Exception();
+        }
+        catch (HBCI_Exception he)
+        {
+          throw he;
+        }
+        catch (Exception e)
+        {
+        	throw new HBCI_Exception(e);
         }
     }
     
Index: src/org/kapott/hbci/manager/HBCIUser.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/manager/HBCIUser.java,v
retrieving revision 1.1
diff -u -r1.1 HBCIUser.java
--- src/org/kapott/hbci/manager/HBCIUser.java	4 May 2011 22:37:46 -0000	1.1
+++ src/org/kapott/hbci/manager/HBCIUser.java	31 Aug 2011 14:03:33 -0000
@@ -574,7 +574,7 @@
     {
         try {
             HBCIUtilsInternal.getCallback().status(passport,HBCICallback.STATUS_INIT_UPD,null);
-            HBCIUtils.log("fetching UPD",HBCIUtils.LOG_INFO);
+            HBCIUtils.log("fetching UPD (BPD-Version: " + passport.getBPDVersion() + ")",HBCIUtils.LOG_INFO);
             
             // autosecmech
             HBCIUtils.log("checking whether passport is supported (but ignoring result)",HBCIUtils.LOG_DEBUG);
@@ -649,9 +649,13 @@
         }
         
         Properties upd=passport.getUPD();
+        Properties bpd=passport.getBPD();
         String     hbciVersionOfUPD=(upd!=null)?upd.getProperty("_hbciversion"):null;
-        
-        if (passport.getUPDVersion().equals("0") ||
+
+        // Wir haben noch keine BPD. Offensichtlich unterstuetzt die Bank
+        // das Abrufen von BPDs ueber einen anonymen Dialog nicht. Also machen
+        // wir das jetzt hier mit einem nicht-anonymen Dialog gleich mit
+        if (bpd == null || passport.getUPDVersion().equals("0") ||
             hbciVersionOfUPD==null ||
             !hbciVersionOfUPD.equals(kernel.getHBCIVersion())) 
         {
