Index: src/org/kapott/hbci/passport/AbstractHBCIPassport.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/AbstractHBCIPassport.java,v
retrieving revision 1.1
diff -u -r1.1 AbstractHBCIPassport.java
--- src/org/kapott/hbci/passport/AbstractHBCIPassport.java	4 May 2011 22:37:43 -0000	1.1
+++ src/org/kapott/hbci/passport/AbstractHBCIPassport.java	24 Nov 2011 21:48:24 -0000
@@ -1,5 +1,5 @@
 
-/*  $Id: 33-catch-exception.patch,v 1.1 2011/11/24 21:57:14 willuhn Exp $
+/*  $Id: 33-catch-exception.patch,v 1.1 2011/11/24 21:57:14 willuhn Exp $
 
     This file is part of HBCI4Java
     Copyright (C) 2001-2008  Stefan Palme
@@ -23,6 +23,7 @@
 
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Enumeration;
@@ -704,10 +705,26 @@
             Constructor con=cl.getConstructor(new Class[] {Object.class});
             HBCIPassport p=(HBCIPassport)(con.newInstance(new Object[] {init}));
             return p;
-        } catch (ClassNotFoundException e) {
-            throw new InvalidUserDataException("*** No passport implementation '"+name+"' found - there must be a class "+className);
-        } catch (Exception e) {
-            throw new HBCI_Exception(HBCIUtilsInternal.getLocMsg("EXCMSG_PASSPORT_INST",name),e); 
+        }
+        catch (ClassNotFoundException e)
+        {
+          throw new InvalidUserDataException("*** No passport implementation '"+name+"' found - there must be a class "+className);
+        }
+        catch (InvocationTargetException ite)
+        {
+          Throwable cause = ite.getCause();
+          if (cause instanceof HBCI_Exception)
+            throw (HBCI_Exception) cause;
+          
+          throw new HBCI_Exception(HBCIUtilsInternal.getLocMsg("EXCMSG_PASSPORT_INST",name),ite);
+        }
+        catch (HBCI_Exception he)
+        {
+          throw he;
+        }
+        catch (Exception e)
+        {
+          throw new HBCI_Exception(HBCIUtilsInternal.getLocMsg("EXCMSG_PASSPORT_INST",name),e); 
         }
     }
 
