Index: src/org/kapott/hbci/passport/AbstractHBCIPassport.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/AbstractHBCIPassport.java,v
retrieving revision 1.2
diff -u -r1.2 AbstractHBCIPassport.java
--- src/org/kapott/hbci/passport/AbstractHBCIPassport.java	24 Nov 2011 21:57:14 -0000	1.2
+++ src/org/kapott/hbci/passport/AbstractHBCIPassport.java	4 Mar 2012 12:40:16 -0000
@@ -948,7 +948,7 @@
         byte[] result=buffer;
         
         if (buffer.length!=size) {
-            HBCIUtils.log("checking for crypted_data_length=="+size+"; current length is "+buffer.length,HBCIUtils.LOG_WARN);
+            HBCIUtils.log("checking for crypted_data_length=="+size+"; current length is "+buffer.length,HBCIUtils.LOG_DEBUG);
             if (buffer.length>size) {
                 int diff=buffer.length-size;
                 boolean ok=true;
@@ -961,7 +961,7 @@
                 }
 
                 if (ok) {
-                    HBCIUtils.log("removing "+diff+" unnecessary null-bytes from crypted_data",HBCIUtils.LOG_WARN);
+                    HBCIUtils.log("removing "+diff+" unnecessary null-bytes from crypted_data",HBCIUtils.LOG_DEBUG);
                     result=new byte[size];
                     System.arraycopy(buffer,diff,result,0,size);
                 }
