Index: src/org/kapott/hbci/passport/AbstractHBCIPassport.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/AbstractHBCIPassport.java,v
retrieving revision 1.3
diff -u -r1.3 AbstractHBCIPassport.java
--- src/org/kapott/hbci/passport/AbstractHBCIPassport.java	4 Mar 2012 12:41:01 -0000	1.3
+++ src/org/kapott/hbci/passport/AbstractHBCIPassport.java	13 Mar 2012 22:07:15 -0000
@@ -137,7 +137,7 @@
 
         if (needPort && 
             (getPort()==null || getPort().intValue()==0)) {
-            StringBuffer sb=new StringBuffer("3000");
+            StringBuffer sb=new StringBuffer((this instanceof AbstractPinTanPassport) ? "443" : "3000");
             HBCIUtilsInternal.getCallback().callback(this,HBCICallback.NEED_PORT,HBCIUtilsInternal.getLocMsg("PORT"),HBCICallback.TYPE_TEXT,sb);
             if (sb.length()==0)
                 throw new InvalidUserDataException(HBCIUtilsInternal.getLocMsg("EXCMSG_EMPTY_X",HBCIUtilsInternal.getLocMsg("PORT")));
Index: src/org/kapott/hbci/passport/HBCIPassportPinTan.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/passport/HBCIPassportPinTan.java,v
retrieving revision 1.5
diff -u -r1.5 HBCIPassportPinTan.java
--- src/org/kapott/hbci/passport/HBCIPassportPinTan.java	30 May 2011 12:47:56 -0000	1.5
+++ src/org/kapott/hbci/passport/HBCIPassportPinTan.java	13 Mar 2012 22:07:15 -0000
@@ -1,5 +1,5 @@
 
-/*  $Id: 38-bug1207.patch,v 1.1 2012/03/13 22:07:43 willuhn Exp $
+/*  $Id: 38-bug1207.patch,v 1.1 2012/03/13 22:07:43 willuhn Exp $
 
     This file is part of HBCI4Java
     Copyright (C) 2001-2008  Stefan Palme
@@ -97,7 +97,6 @@
         
         HBCIUtils.log("loading passport data from file "+fname,HBCIUtils.LOG_DEBUG);
         setFileName(fname);
-        setPort(new Integer(443));
         setCertFile(HBCIUtils.getParam(header+"certfile"));
         setCheckCert(HBCIUtils.getParam(header+"checkcert","1").equals("1"));
         
