Index: src/org/kapott/hbci/rewrite/RSigIdLeadingZero.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/rewrite/RSigIdLeadingZero.java,v
retrieving revision 1.1
diff -u -r1.1 RSigIdLeadingZero.java
--- src/org/kapott/hbci/rewrite/RSigIdLeadingZero.java	4 May 2011 22:37:57 -0000	1.1
+++ src/org/kapott/hbci/rewrite/RSigIdLeadingZero.java	27 Mar 2012 21:31:58 -0000
@@ -1,5 +1,5 @@
 
-/*  $Id: 39-blz-optional.patch,v 1.1 2012/03/27 21:33:13 willuhn Exp $
+/*  $Id: 39-blz-optional.patch,v 1.1 2012/03/27 21:33:13 willuhn Exp $
 
     This file is part of HBCI4Java
     Copyright (C) 2001-2008  Stefan Palme
@@ -46,9 +46,9 @@
                 if (secondPlus!=-1) {
                     StringBuffer value=new StringBuffer(st.substring(firstPlus+1,secondPlus));
                     
-                    if (value.length()>0 && value.charAt(0)=='0') {
+                    if (value.length()>1 && value.charAt(0)=='0') {
                         HBCIUtils.log("RSigIdLeadingZero: found leading zero ("+value+"), removing it",HBCIUtils.LOG_WARN);
-                        while (value.length()>0 && value.charAt(0)=='0') {
+                        while (value.length()>1 && value.charAt(0)=='0') {
                             value.deleteCharAt(0);
                         }
                         
Index: src/org/kapott/hbci/security/Sig.java
===================================================================
RCS file: /cvsroot/hibiscus/hbci4java/src/org/kapott/hbci/security/Sig.java,v
retrieving revision 1.1
diff -u -r1.1 Sig.java
--- src/org/kapott/hbci/security/Sig.java	4 May 2011 22:38:03 -0000	1.1
+++ src/org/kapott/hbci/security/Sig.java	27 Mar 2012 21:31:58 -0000
@@ -1,5 +1,5 @@
 
-/*  $Id: 39-blz-optional.patch,v 1.1 2012/03/27 21:33:13 willuhn Exp $
+/*  $Id: 39-blz-optional.patch,v 1.1 2012/03/27 21:33:13 willuhn Exp $
 
     This file is part of HBCI4Java
     Copyright (C) 2001-2008  Stefan Palme
@@ -409,7 +409,6 @@
 
         u_role = msg.getValueOfDE(sigheadName + ".role");
         u_range = msg.getValueOfDE(sigheadName + ".range");
-        u_keyblz = msg.getValueOfDE(sigheadName + ".KeyName.KIK.blz");
         u_keycountry = msg.getValueOfDE(sigheadName + ".KeyName.KIK.country");
         u_keyuserid = msg.getValueOfDE(sigheadName + ".KeyName.userid");
         u_keynum = msg.getValueOfDE(sigheadName + ".KeyName.keynum");
@@ -418,7 +417,16 @@
         u_sigalg = msg.getValueOfDE(sigheadName + ".SigAlg.alg");
         u_sigmode = msg.getValueOfDE(sigheadName + ".SigAlg.mode");
         u_hashalg = msg.getValueOfDE(sigheadName + ".HashAlg.alg");
-        
+
+        // Die Angabe der BLZ ist nicht unbedingt verpflichtend (für 280 aber schon...). Trotzdem gibt es wohl
+        // Banken die das nicht interessiert...
+        try {
+          u_keyblz = msg.getValueOfDE(sigheadName + ".KeyName.KIK.blz");
+        }
+        catch (Exception e) {
+          HBCIUtils.log("missing bank code in message signature, ignoring...",HBCIUtils.LOG_WARN);
+        }
+
         if (mainPassport.needUserSig()) {
             // TODO: bei anderen user-signaturen hier allgemeineren code schreiben
             Hashtable values=new Hashtable();
