diff --git a/src/org/kapott/hbci/manager/HBCIUtils.java b/src/org/kapott/hbci/manager/HBCIUtils.java
index 92103c0..1f1abd4 100644
--- a/src/org/kapott/hbci/manager/HBCIUtils.java
+++ b/src/org/kapott/hbci/manager/HBCIUtils.java
@@ -915,7 +915,12 @@ public final class HBCIUtils
       // Aber nur, wenn sie numerisch ist. Bei irgendeiner Bank wurde
       // "EUR" als Unterkontonummer verwendet. Das geht natuerlich nicht,
       // weil damit nicht gerechnet werden kann
-      if (k.subnumber != null && k.subnumber.length() > 0 && k.subnumber.matches("[0-9]{1,8}"));
+      // Wir machen das auch nur dann, wenn beide Nummern zusammen max.
+      // 10 Zeichen ergeben
+      if (k.subnumber != null &&
+          k.subnumber.length() > 0 &&
+          k.subnumber.matches("[0-9]{1,8}") &&
+          k.number.length() + k.subnumber.length() <= 10)
         konto += k.subnumber;
       
       /////////////////
